subprojects/build-cache-http/src/main/java/org/gradle/caching/http/internal/HttpBuildCacheService.java

```diff
@@ -95,45 +97,27 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) throws Buil
                 return false;
             } else {
                 String defaultMessage = String.format("Loading entry from '%s' response status %d: %s", safeUri(uri), statusCode, statusLine.getReasonPhrase());
-                if (isRedirect(statusCode)) {
-                    return handleRedirect(uri, response, statusCode, defaultMessage, "loading entry from");
-                } else {
-                    return throwHttpStatusCodeException(statusCode, defaultMessage);
-                }
+                return throwHttpStatusCodeException(statusCode, defaultMessage);
             }
         } catch (IOException e) {
             throw wrap(e);
         }
     }
 
-    private boolean handleRedirect(URI uri, HttpClientResponse response, int statusCode, String defaultMessage, String action) {
-        String locationHeader = response.getHeader(HttpHeaders.LOCATION);
-        if (locationHeader == null) {
-            return throwHttpStatusCodeException(statusCode, defaultMessage);
-        }
-        try {
-            throw new BuildCacheException(String.format("Received unexpected redirect (HTTP %d) to %s when " + action + " '%s'. "
-                + "Ensure the configured URL for the remote build cache is correct.", statusCode, safeUri(new URI(locationHeader)), safeUri(uri)));
-        } catch (URISyntaxException e) {
-            return throwHttpStatusCodeException(statusCode, defaultMessage);
-        }
-    }
-
-    private boolean isRedirect(int statusCode) {
-        return statusCode == HttpStatus.SC_MOVED_PERMANENTLY || statusCode == HttpStatus.SC_MOVED_TEMPORARILY || statusCode == HttpStatus.SC_TEMPORARY_REDIRECT;
-    }
-
     @Override
     public void store(BuildCacheKey key, final BuildCacheEntryWriter output) throws BuildCacheException {
         final URI uri = root.resolve(key.getHashCode());
         HttpPut httpPut = new HttpPut(uri);
+        if (useExpectContinue) {
+            httpPut.setHeader(HTTP.EXPECT_DIRECTIVE, HTTP.EXPECT_CONTINUE);
+        }
         httpPut.addHeader(HttpHeaders.CONTENT_TYPE, BUILD_CACHE_CONTENT_TYPE);
         requestCustomizer.customize(httpPut);
 
         httpPut.setEntity(new AbstractHttpEntity() {
             @Override
             public boolean isRepeatable() {
-                return false;
+                return true;
```

Comment:
It's always repeatable, because we are always reading from a file. We can just open up another stream.

Repeating the file transmission is potentially expensive, which is you would want to use expect-continue. However, expect-continue isn't reliable with all servers and does impose a small performance penalty so I don't think turning it on by default is the right thing to do (in case that was to be a follow up question).